[UNOMI-638] - Updates documentation in preparation for Unomi 2 (#472)
apache/unomi: manual/src/main/asciidoc/migrations/migrate-1.6-to-2.0.adoc

==== Data Model changes

There has been changes to Unomi Data model, please make sure to review those in the << what_s_new>> section of the user manual.

==== Create JSON schemas

Once you updated your applications to align with Unomi 2 data model, the next step will be to create the necessary JSON Schemas.

Any event (and more generally, any object) received through Unomi public endpoints do require a valid JSON schema. 
Apache Unomi ships, out of the box, with all of the necessary JSON Schemas for its own operation but you will need to create schemas for any custom event. 

When creating your new schemas, reviewing debug messages in the logs (using: `log:set DEBUG org.apache.unomi.schema.impl.SchemaServiceImpl` in Karaf console), 
will point to errors in your schemas.

Note that it is currently not possible to modify or surcharge an existing JSON schema via the REST API. 
If you are using custom properties on an event type for which an Apache Unomi schema already exists, 
you will need to either change its eventType and create the corresponding Schema or to create a Unomi extension.

> You can use as a source of inspiration for creating new schemas, Apache Unomi 2.0 schema located at: 

Reviewer: ```suggestion
You can use, as a source of inspiration for creating new schemas, Apache Unomi 2.0 schema located at: 
```

 https://github.com/apache/unomi/tree/master/extensions/json-schema/services/src/main/resources/META-INF/cxs/schemas[extensions/json-schema/services/src/main/resources/META-INF/cxs/schemas].

> Finally, and altuough it is technically feasible, we recommend against creating permissive JSON Schemas allowing any event payload.

Reviewer: ```suggestion
Finally, and although it is technically feasible, we recommend against creating permissive JSON Schemas allowing any event payload. This requires making sure that you don't allow open properties by using JSON schema keywords such as https://json-schema.org/understanding-json-schema/reference/object.html#unevaluated-properties[unevaluated properties]
```

At a minimum, the required capacity must be greater than the size of the dataset in production + the size of the largest index.

==== Migrate custom data

Apache Unomi 2.0 knows how to migrate its own data from the new model to the old one, but it does not know how to migrate custom events you might be using in your environment.

It relies on a set of groovy scripts to perform its data migration, 
located in https://github.com/apache/unomi/tree/master/tools/shell-commands/src/main/resources/META-INF/cxs/migration[tools/shell-commands/src/main/resources/META-INF/cxs/migration], 
these scripts are sorted alphabetically and executed sequentially when migration is started. You can use these scripts as a source of inspiration for creating your own.

In most cases, migration steps consist of an Elasticsearch painless script that will handle the data changes.

At runtime, and when starting the migration, Unomi 2.0 will take its own scripts, any additional scripts located in `data/migration/scripts`, will sort the resulting list alphabetically and execute each migration script sequentially.

==== Perform the migration

===== Checklist

> Before starting the mirgation, please ensure that:

Reviewer: ```suggestion
Before starting the migration, please ensure that:
```

===== Configuration

The following environment variables are used for the migration:

|===
|Environment Variable|Unomi Setting|Default

|UNOMI_ELASTICSEARCH_ADDRESSES
|org.apache.unomi.elasticsearch.addresses
|localhost:9200

|UNOMI_ELASTICSEARCH_SSL_ENABLE
|org.apache.unomi.elasticsearch.sslEnable
|false

|UNOMI_ELASTICSEARCH_USERNAME
|org.apache.unomi.elasticsearch.username
|

|UNOMI_ELASTICSEARCH_PASSWORD
|org.apache.unomi.elasticsearch.password
|

|UNOMI_ELASTICSEARCH_SSL_TRUST_ALL_CERTIFICATES
|org.apache.unomi.elasticsearch.sslTrustAllCertificates
|false

|UNOMI_ELASTICSEARCH_INDEXPREFIX
|org.apache.unomi.elasticsearch.index.prefix
|context

|UNOMI_MIGRATION_RECOVER_FROM_HISTORY
|org.apache.unomi.migration.recoverFromHistory
|true

|===

If there is a need for advanced configuratiion, the configuration file used by Apache Unomi 2.0 is located in: `etc/org.apache.unomi.migration.cfg`

===== Migrate manually

You can migrate manually using the Karaf console. 

After having started Apache Unomi 2.0 with the `./karaf` command, you will be presented with the Karaf shell.

From there you have two options:

>  - The necessary configuration variables (see above) are set, the migration can be started using: `unomi:migrate 1.6.0`

Reviewer: ```suggestion
 - The necessary configuration variables (see above) have already been set, you can start the migration using the command: `unomi:migrate 1.6.0`
```

>  - Or, you want to provide the configuration settings via the terminal, in that case you can start the migration in interactive mode using: `unomi:migrate 1.6.0`

Reviewer: ```suggestion
 - Or, you want to provide the configuration settings interactively via the terminal, in that case you can start the migration in interactive mode using: `unomi:migrate 1.6.0`
```

> The value of the migrate command (1.6.0 in the example above), corresponds to the version you're migrating from.

Reviewer: ```suggestion
The parameter of the migrate command (1.6.0 in the example above) corresponds to the version you're migrating from.
```

At the end of the migration, you can start Unomi 2.0 as usual using: `unomi:start`.

===== Migrate with Docker

The migration can also be performed using Docker images, the migration itself can be started by passing a specific value to the `KARAF_OPTS` environment variable.

In the context of this migration guide, we will asssume that: 
 - Custom migration scripts are located in `/home/unomi/migration/scripts/`
 - Painless scripts, or more generally any migration assets are located in `/home/unomi/migration/assets/`, these scripts will be mounted under `/tmp/assets/` inside the Docker container. 

[source]
----
docker run \
    -e UNOMI_ELASTICSEARCH_ADDRESSES=localhost:9200 \
    -e KARAF_OPTS="-Dunomi.autoMigrate=1.6.0" \
    --v /home/unomi/migration/scripts/:/opt/apache-unomi/data/migration/scripts \
    --v /home/unomi/migration/assets/:/tmp/assets/ \
    apache/unomi:2.0.0-SNAPSHOT
----

You might need to provide additional variables (see table above) depending of your environment.

> If the migration fail, you can simply restart this command.

Reviewer: ```suggestion
If the migration fails, you can simply restart this command.
```
